{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "be16745e",
   "metadata": {},
   "source": [
    "## Load necessary python libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1a0b5448",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_7465/2314815770.py:4: DeprecationWarning: \n",
      "Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),\n",
      "(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)\n",
      "but was not found to be installed on your system.\n",
      "If this would cause problems for you,\n",
      "please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466\n",
      "        \n",
      "  import pandas as pd\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Big-FISH version: 0.6.2\n",
      "numpy version: 1.23.0\n",
      "Average measured size from image : 120.5078125\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import napari\n",
    "import tifffile\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from copy import deepcopy\n",
    "import bigfish\n",
    "import bigfish.stack as stack\n",
    "import bigfish.detection as detection\n",
    "import bigfish.multistack as multistack\n",
    "\n",
    "import bigfish.plot as plot\n",
    "\n",
    "from buildReferenceSpot import buildReferenceSpotFromImages\n",
    "from runBigfishDetection import saveSpotsNPZ\n",
    "from bigfish.detection.utils import get_object_radius_pixel\n",
    "from bigfish.detection.utils import get_breaking_point\n",
    "from matplotlib.backends.backend_qt5agg import FigureCanvas\n",
    "from dask.array.image import imread as imr\n",
    "from imaris_ims_file_reader.ims import ims\n",
    "from kneed import KneeLocator\n",
    "\n",
    "from pathlib import Path\n",
    "from napari.layers import Image\n",
    "from magicgui import magicgui, magic_factory, widgets\n",
    "from magicgui.widgets import Table\n",
    "\n",
    "print(\"Big-FISH version: {0}\".format(bigfish.__version__))\n",
    "print(\"numpy version: {0}\".format(np.__version__))\n",
    "\n",
    "print('Average measured size from image :',(123.4/1024)*1*1000) #objectRad in nm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69073c90",
   "metadata": {},
   "source": [
    "# Run GUI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0fbfc931",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<napari._qt.widgets.qt_viewer_dock_widget.QtViewerDockWidget at 0x7f33efda7040>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no cells found\n",
      "calling getFolderName\n",
      "calling getFolderName\n",
      "calling getFolderName\n",
      "calling getFolderName\n",
      "calling showMovie\n",
      "Stackcell found\n",
      "getting mip\n",
      "no cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/components/viewer_model.py\", line 1395, in _add_layer_from_data\n",
      "    layer = add_method(data, **(meta or {}))\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/utils/migrations.py\", line 62, in _update_from_dict\n",
      "    return func(*args, **kwargs)\n",
      "TypeError: add_image() got an unexpected keyword argument 'cellFolderName'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 418, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 256, in _on_change\n",
      "    self()\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 354, in __call__\n",
      "    callback(self, value, return_type)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/utils/_magicgui.py\", line 114, in add_layer_data_tuples_to_viewer\n",
      "    _add_layer_data_tuples_to_viewer(\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/_app_model/injection/_processors.py\", line 35, in _add_layer_data_tuples_to_viewer\n",
      "    viewer._add_layer_from_data(*datum)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/components/viewer_model.py\", line 1400, in _add_layer_from_data\n",
      "    raise TypeError(\n",
      "TypeError: _add_layer_from_data received an unexpected keyword argument ('cellFolderName') for layer type image\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 321, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_container_widget.py\", line 213, in <lambda>\n",
      "    widget.changed.connect(lambda: self.changed.emit(self))\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets._function_gui.FunctionGui.changed', an error occurred in callback 'magicgui.widgets._function_gui.FunctionGui._on_change'.\n",
      "The args passed to the callback were: (<FunctionGui showMovie(input: str = '/media/raphael/data/works/20231220_C3_14_2h_ofON/bleach_correction/expo/C3.14_ON_01_w1SPI-488_s4/cell_26', showMip: bool = True) -> <function NewType.<locals>.new_type at 0x7f3404de9280>>,)\n",
      "This is not a bug in psygnal.  See 'TypeError' above for details.\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_value_widget.py\", line 71, in _on_value_change\n",
      "    self.changed.emit(value)\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets.LineEdit.changed', an error occurred in callback 'magicgui.widgets.bases._container_widget.ContainerWidget.insert.<locals>.<lambda>'.\n",
      "The args passed to the callback were: ('/media/raphael/data/works/20231220_C3_14_2h_ofON/bleach_correction/expo/C3.14_ON_01_w1SPI-488_s4/cell_26',)\n",
      "This is not a bug in psygnal.  See 'EmitLoopError' above for details.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "calling showMovie\n",
      "Stackcell found\n",
      "no cells found\n",
      "no cells found\n",
      "no cells found\n",
      "calling getFolderName\n",
      "calling getFolderName\n",
      "calling showMovie\n",
      "Stackcell found\n",
      "getting mip\n",
      "no cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/components/viewer_model.py\", line 1395, in _add_layer_from_data\n",
      "    layer = add_method(data, **(meta or {}))\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/utils/migrations.py\", line 62, in _update_from_dict\n",
      "    return func(*args, **kwargs)\n",
      "TypeError: add_image() got an unexpected keyword argument 'cellFolderName'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 418, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 256, in _on_change\n",
      "    self()\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 354, in __call__\n",
      "    callback(self, value, return_type)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/utils/_magicgui.py\", line 114, in add_layer_data_tuples_to_viewer\n",
      "    _add_layer_data_tuples_to_viewer(\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/_app_model/injection/_processors.py\", line 35, in _add_layer_data_tuples_to_viewer\n",
      "    viewer._add_layer_from_data(*datum)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/components/viewer_model.py\", line 1400, in _add_layer_from_data\n",
      "    raise TypeError(\n",
      "TypeError: _add_layer_from_data received an unexpected keyword argument ('cellFolderName') for layer type image\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 321, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_container_widget.py\", line 213, in <lambda>\n",
      "    widget.changed.connect(lambda: self.changed.emit(self))\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets._function_gui.FunctionGui.changed', an error occurred in callback 'magicgui.widgets._function_gui.FunctionGui._on_change'.\n",
      "The args passed to the callback were: (<FunctionGui showMovie(input: str = '/media/raphael/data/works/20231220_C3_14_2h_ofON/bleach_correction/expo/C3.14_ON_01_w1SPI-488_s4/cell_26', showMip: bool = True) -> <function NewType.<locals>.new_type at 0x7f3404de9280>>,)\n",
      "This is not a bug in psygnal.  See 'TypeError' above for details.\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_value_widget.py\", line 71, in _on_value_change\n",
      "    self.changed.emit(value)\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets.CheckBox.changed', an error occurred in callback 'magicgui.widgets.bases._container_widget.ContainerWidget.insert.<locals>.<lambda>'.\n",
      "The args passed to the callback were: (True,)\n",
      "This is not a bug in psygnal.  See 'EmitLoopError' above for details.\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 321, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 224, in _disable_button_and_call\n",
      "    self.__call__()\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 342, in __call__\n",
      "    value = self._function(*bound.args, **bound.kwargs)\n",
      "  File \"/tmp/ipykernel_7465/77003041.py\", line 239, in recordThreshold\n",
      "    stackCellSample = image_data[::samples,:,:,:]\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/dask/array/core.py\", line 1979, in __getitem__\n",
      "    index2 = normalize_index(index, self.shape)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/dask/array/slicing.py\", line 909, in normalize_index\n",
      "    raise IndexError(\"Too many indices for array\")\n",
      "IndexError: Too many indices for array\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_value_widget.py\", line 71, in _on_value_change\n",
      "    self.changed.emit(value)\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets.PushButton.changed', an error occurred in callback 'magicgui.widgets._function_gui.FunctionGui.__init__.<locals>._disable_button_and_call'.\n",
      "The args passed to the callback were: (False,)\n",
      "This is not a bug in psygnal.  See 'IndexError' above for details.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "calling getThreshold\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/components/viewer_model.py\", line 1395, in _add_layer_from_data\n",
      "    layer = add_method(data, **(meta or {}))\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/utils/migrations.py\", line 62, in _update_from_dict\n",
      "    return func(*args, **kwargs)\n",
      "TypeError: add_image() got an unexpected keyword argument 'samplingRate'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 321, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 224, in _disable_button_and_call\n",
      "    self.__call__()\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 354, in __call__\n",
      "    callback(self, value, return_type)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/utils/_magicgui.py\", line 114, in add_layer_data_tuples_to_viewer\n",
      "    _add_layer_data_tuples_to_viewer(\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/_app_model/injection/_processors.py\", line 35, in _add_layer_data_tuples_to_viewer\n",
      "    viewer._add_layer_from_data(*datum)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/components/viewer_model.py\", line 1400, in _add_layer_from_data\n",
      "    raise TypeError(\n",
      "TypeError: _add_layer_from_data received an unexpected keyword argument ('samplingRate') for layer type image\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_value_widget.py\", line 71, in _on_value_change\n",
      "    self.changed.emit(value)\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets.PushButton.changed', an error occurred in callback 'magicgui.widgets._function_gui.FunctionGui.__init__.<locals>._disable_button_and_call'.\n",
      "The args passed to the callback were: (False,)\n",
      "This is not a bug in psygnal.  See 'TypeError' above for details.\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 321, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 224, in _disable_button_and_call\n",
      "    self.__call__()\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 342, in __call__\n",
      "    value = self._function(*bound.args, **bound.kwargs)\n",
      "  File \"/tmp/ipykernel_7465/77003041.py\", line 239, in recordThreshold\n",
      "    stackCellSample = image_data[::samples,:,:,:]\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/dask/array/core.py\", line 1979, in __getitem__\n",
      "    index2 = normalize_index(index, self.shape)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/dask/array/slicing.py\", line 909, in normalize_index\n",
      "    raise IndexError(\"Too many indices for array\")\n",
      "IndexError: Too many indices for array\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_value_widget.py\", line 71, in _on_value_change\n",
      "    self.changed.emit(value)\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets.PushButton.changed', an error occurred in callback 'magicgui.widgets._function_gui.FunctionGui.__init__.<locals>._disable_button_and_call'.\n",
      "The args passed to the callback were: (False,)\n",
      "This is not a bug in psygnal.  See 'IndexError' above for details.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no cells found\n",
      "calling getFolderName\n",
      "calling showMovie\n",
      "Stackcell found\n",
      "getting mip\n",
      "no cells found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/components/viewer_model.py\", line 1395, in _add_layer_from_data\n",
      "    layer = add_method(data, **(meta or {}))\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/utils/migrations.py\", line 62, in _update_from_dict\n",
      "    return func(*args, **kwargs)\n",
      "TypeError: add_image() got an unexpected keyword argument 'cellFolderName'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 418, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 256, in _on_change\n",
      "    self()\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 354, in __call__\n",
      "    callback(self, value, return_type)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/utils/_magicgui.py\", line 114, in add_layer_data_tuples_to_viewer\n",
      "    _add_layer_data_tuples_to_viewer(\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/_app_model/injection/_processors.py\", line 35, in _add_layer_data_tuples_to_viewer\n",
      "    viewer._add_layer_from_data(*datum)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/napari/components/viewer_model.py\", line 1400, in _add_layer_from_data\n",
      "    raise TypeError(\n",
      "TypeError: _add_layer_from_data received an unexpected keyword argument ('cellFolderName') for layer type image\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 321, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_container_widget.py\", line 213, in <lambda>\n",
      "    widget.changed.connect(lambda: self.changed.emit(self))\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets._function_gui.FunctionGui.changed', an error occurred in callback 'magicgui.widgets._function_gui.FunctionGui._on_change'.\n",
      "The args passed to the callback were: (<FunctionGui showMovie(input: str = '/media/raphael/data/works/20231220_C3_14_2h_ofON/bleach_correction/expo/C3.14_ON_01_w1SPI-488_s4/cell_41', showMip: bool = True) -> <function NewType.<locals>.new_type at 0x7f3404de9280>>,)\n",
      "This is not a bug in psygnal.  See 'TypeError' above for details.\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_value_widget.py\", line 71, in _on_value_change\n",
      "    self.changed.emit(value)\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets.LineEdit.changed', an error occurred in callback 'magicgui.widgets.bases._container_widget.ContainerWidget.insert.<locals>.<lambda>'.\n",
      "The args passed to the callback were: ('/media/raphael/data/works/20231220_C3_14_2h_ofON/bleach_correction/expo/C3.14_ON_01_w1SPI-488_s4/cell_41',)\n",
      "This is not a bug in psygnal.  See 'EmitLoopError' above for details.\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"src/psygnal/_signal.py\", line 990, in _run_emit_loop\n",
      "  File \"src/psygnal/_weak_callback.py\", line 321, in cb\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 224, in _disable_button_and_call\n",
      "    self.__call__()\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/_function_gui.py\", line 342, in __call__\n",
      "    value = self._function(*bound.args, **bound.kwargs)\n",
      "  File \"/tmp/ipykernel_7465/77003041.py\", line 239, in recordThreshold\n",
      "    stackCellSample = image_data[::samples,:,:,:]\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/dask/array/core.py\", line 1979, in __getitem__\n",
      "    index2 = normalize_index(index, self.shape)\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/dask/array/slicing.py\", line 909, in normalize_index\n",
      "    raise IndexError(\"Too many indices for array\")\n",
      "IndexError: Too many indices for array\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/raphael/miniconda3/envs/bigfishlive/lib/python3.9/site-packages/magicgui/widgets/bases/_value_widget.py\", line 71, in _on_value_change\n",
      "    self.changed.emit(value)\n",
      "  File \"src/psygnal/_signal.py\", line 945, in emit\n",
      "  File \"src/psygnal/_signal.py\", line 992, in _run_emit_loop\n",
      "psygnal._exceptions.EmitLoopError: \n",
      "While emitting signal 'magicgui.widgets.PushButton.changed', an error occurred in callback 'magicgui.widgets._function_gui.FunctionGui.__init__.<locals>._disable_button_and_call'.\n",
      "The args passed to the callback were: (False,)\n",
      "This is not a bug in psygnal.  See 'IndexError' above for details.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "VOXEL_RADIUS = (0,0,0)\n",
    "OBJECT_RADIUS = (0,0,0)\n",
    "\n",
    "DEFAULT_CHOICES=['None']\n",
    "SET_TABLE = 0\n",
    "CELL_CHOICE = 0\n",
    "DEBUG = False\n",
    "@magicgui(auto_call=True,    \n",
    "    main_window=False,\n",
    "          persist=False,\n",
    "    base_folder={\"label\": \"Choose home folder:\",\n",
    "                 'mode': 'd'},\n",
    "    dropdown=dict(\n",
    "            widget_type=\"Select\", choices=DEFAULT_CHOICES, label=\"Choose cell to analyse\"\n",
    "        ),\n",
    "    layout=\"vertical\",\n",
    "         )\n",
    "def choose_home_folder(dropdown, base_folder = Path.home()):\n",
    "    if DEBUG==True:    \n",
    "        print(\"calling choose home folder\")\n",
    "        print('dropdown :', dropdown)\n",
    "        print('bf: ', base_folder)\n",
    "\n",
    "    return base_folder, dropdown\n",
    "\n",
    "@choose_home_folder.changed.connect\n",
    "def choose_cell_on_file_change(event=None):\n",
    "    global SET_TABLE, CELL_CHOICE\n",
    "    if DEBUG==True: \n",
    "        print('calling populate cell choices')\n",
    "        print('choose home folder count :', choose_home_folder.call_count)\n",
    "    if choose_home_folder.call_count == 1 and SET_TABLE==0:\n",
    "        SET_TABLE = 1\n",
    "        if DEBUG==True: print('value of set table set to : ', SET_TABLE)\n",
    "    hm = choose_home_folder.base_folder.value\n",
    "    cellChoices = [i for i in os.listdir(hm) if 'cell_' in i]\n",
    "    if cellChoices is not None:\n",
    "        choose_home_folder.dropdown.choices = cellChoices\n",
    "        global DEFAULT_CHOICES\n",
    "        DEFAULT_CHOICES = cellChoices\n",
    "        choose_home_folder.dropdown.value = cellChoices[CELL_CHOICE]\n",
    "        \n",
    "\n",
    "\n",
    "@choose_home_folder.called.connect\n",
    "def on_dropdown_changed(value: str):\n",
    "    def catchval(new_value):\n",
    "        thresholdTable[new_value['column_header']][new_value['row_header']] = new_value['data']\n",
    "        table.to_dataframe().to_csv(str(value[0])+'/'+'thresholds.csv')\n",
    "    \n",
    "    if DEBUG==True: \n",
    "        print('calling on drowpdown changed')\n",
    "        print(value)\n",
    "    global SET_TABLE, CELL_CHOICE, DEFAULT_CHOICES\n",
    "    if np.size(value[1])!=0:\n",
    "        if DEBUG==True: print('chosen cell', str(value[1][0]))\n",
    "        CELL_CHOICE = DEFAULT_CHOICES.index(value[1][0])\n",
    "        getFolderName.folder_name.value = os.path.join(str(value[0]), str(value[1][0]))\n",
    "        if SET_TABLE==1:\n",
    "            \n",
    "            hm = str(choose_home_folder.base_folder.value)\n",
    "            if os.path.exists(hm+'/'+'thresholds.csv'):\n",
    "                df = pd.read_csv(fn, index_col=0)\n",
    "                thresholdTable = df.to_dict()\n",
    "            else:\n",
    "                if DEBUG==True: print('calling set table', hm)\n",
    "                cellChoices =  [i for i in os.listdir(hm) if 'cell_' in i]\n",
    "                rows = np.arange(len(cellChoices))\n",
    "\n",
    "                rownames = ['r'+str(element) for element in rows]\n",
    "                column1 = {rownames[i]: cellChoices[i] for i in range(len(cellChoices))}\n",
    "                column2 = {rownames[i]: 0 for i in range(len(cellChoices))}\n",
    "                thresholdTable = {\n",
    "                    \"cell\": column1,\n",
    "                    \"threshold\": column2,\n",
    "\n",
    "                }         \n",
    "            table = Table(value=thresholdTable)\n",
    "            table.changed.connect(catchval)\n",
    "            tablewidget = viewer.window.add_dock_widget(table, tabify=False, name = 'threshold table')\n",
    "            SET_TABLE = 0\n",
    "\n",
    "    else:\n",
    "        print('no cells found')\n",
    "        getFolderName.folder_name.value = str(value[0])   \n",
    "        \n",
    "        \n",
    "@magicgui(\n",
    "    auto_call=False,\n",
    "    layout=\"vertical\",\n",
    "    show_MIP={\n",
    "        \"widget_type\": \"RadioButtons\",\n",
    "        \"orientation\": \"horizontal\",\n",
    "        \"choices\": [(\"yes\", True), (\"no\", False)],\n",
    "    },\n",
    ")\n",
    "def getFolderName(\n",
    "    folder_name=\"cell folder\",\n",
    "    show_MIP=False,\n",
    "    \n",
    "    ):\n",
    "    \"\"\"Callable function A.\"\"\"\n",
    "    print(\"calling getFolderName\")\n",
    "    if not os.path.exists(folder_name):\n",
    "        folder_name = Path.home()\n",
    "    return folder_name, show_MIP\n",
    "\n",
    "@getFolderName.called.connect\n",
    "def _on_func_getFolderName(value: str):\n",
    "    showMovie.input.value = value[0]\n",
    "    showMovie.showMip.value = value[1]\n",
    "    \n",
    "\n",
    "@magicgui(auto_call=True, input={\"visible\": False, \"label\": \" \"}, showMip={\"visible\": False}, persist=True)\n",
    "def showMovie(input: str, showMip :bool)-> \"napari.types.LayerDataTuple\":\n",
    "    \"\"\"Callable function B.\"\"\"\n",
    "    print(\"calling showMovie\")\n",
    "    homefolder = input\n",
    "    if any(fname.endswith(r\".tif\") for fname in os.listdir(homefolder)):\n",
    "        print('Stackcell found')\n",
    "        stackCell = imr(homefolder+'/*t*.tif')\n",
    "        metadata = {'name': 'stackCell', 'colormap': 'green', 'cellFolderName': homefolder}\n",
    "        layer_type = 'image'\n",
    "        viewer = napari.current_viewer()\n",
    "\n",
    "        if showMip == True:\n",
    "            print(\"getting mip\")\n",
    "            maxImageCell = np.max(stackCell, axis=1)\n",
    "            image_layer = viewer.add_image(maxImageCell,\n",
    "                         colormap='green')\n",
    "        else:\n",
    "            image_layer = viewer.add_image(stackCell,\n",
    "                         colormap='green')\n",
    "\n",
    "        return (stackCell, metadata, layer_type)\n",
    "\n",
    "@magicgui(main_window=False,\n",
    "          call_button='Set voxel parameters',\n",
    "          voxel_radius_z = {\"widget_type\": \"IntSlider\", 'max': 1000},\n",
    "          voxel_radius_xy = {\"widget_type\": \"IntSlider\", 'max': 1000},\n",
    "          object_radius_z = {\"widget_type\": \"IntSlider\", 'max': 1000},\n",
    "          object_radius_xy = {\"widget_type\": \"IntSlider\", 'max': 1000})\n",
    "\n",
    "def setVoxelandObjectParameters(voxel_radius_z=600,\n",
    "                                voxel_radius_xy = 121,\n",
    "                                object_radius_z=600,\n",
    "                                object_radius_xy=150,\n",
    "    ):\n",
    "\n",
    "    voxelRadius = (voxel_radius_z, voxel_radius_xy, voxel_radius_xy)\n",
    "    objectRadius = (object_radius_z, object_radius_xy, object_radius_xy)\n",
    "    return voxelRadius, objectRadius\n",
    "@setVoxelandObjectParameters.called.connect\n",
    "def on_set_vb_parameters(value: list):\n",
    "    global VOXEL_RADIUS\n",
    "    global OBJECT_RADIUS\n",
    "    VOXEL_RADIUS = value[0]\n",
    "    OBJECT_RADIUS = value[1]\n",
    "    if DEBUG==True: \n",
    "        print('voxel ', VOXEL_RADIUS)\n",
    "        print('object', OBJECT_RADIUS)\n",
    "\n",
    "\n",
    "\n",
    "@magicgui(main_window=False,\n",
    "             sampling_rate={'min': 5, 'max':50},)\n",
    "def getThreshold(img_layer: \"napari.layers.Image\", sampling_rate: int) -> \"napari.types.LayerDataTuple\":\n",
    "    print('calling getThreshold')\n",
    "    image_data = img_layer.data #\n",
    "    layer_type = 'image'\n",
    "    metadata = {'colormap': 'green','name': 'stackCell', 'samplingRate': sampling_rate} \n",
    "    return image_data, metadata, layer_type\n",
    "\n",
    "@getThreshold.called.connect\n",
    "def runThresholdEstimator(value: int):\n",
    "    global VOXEL_RADIUS\n",
    "    global OBJECT_RADIUS\n",
    "    print('calling runThresholEstimator')\n",
    "    mpl_widget = FigureCanvas(plt.figure(figsize=(5,10)))\n",
    "    images = []\n",
    "    spots_list=[]\n",
    "    stackCell = value[0]\n",
    "    MaxTimePoint = len(value[0])\n",
    "    samplingRate = value[1]['samplingRate']\n",
    "    for t in range(0,MaxTimePoint,samplingRate):\n",
    "        rna = np.array(stackCell[t])\n",
    "        images.append(rna)\n",
    "    print(\"Total number of images : \"+str(len(images)))\n",
    "    viewer = napari.current_viewer()\n",
    "    viewer.window.add_dock_widget(mpl_widget,tabify=True, name = 'elbow plot')\n",
    "    spots_image_list, threshold_image_list = detection.detect_spots(\n",
    "    images=images, \n",
    "    return_threshold=True, \n",
    "    voxel_size=VOXEL_RADIUS,  # in nanometer (one value per dimension zyx)\n",
    "    spot_radius=OBJECT_RADIUS)  # in nanometer (one value per dimension zyx)\n",
    "\n",
    "    thresholds, count_spots, threshold = detection.get_elbow_values(\n",
    "            images,\n",
    "            voxel_size=VOXEL_RADIUS,\n",
    "            spot_radius=OBJECT_RADIUS\n",
    "    )\n",
    "    print('Automatic threshold found!')\n",
    "    knees = []\n",
    "    norm_knees = []\n",
    "\n",
    "    kneedle = KneeLocator(\n",
    "        thresholds, count_spots, S=1, curve=\"convex\", direction=\"decreasing\", online=False)\n",
    "\n",
    "    xx = np.argmax(kneedle.y_difference)\n",
    "    plt.subplot(2, 1, 2)\n",
    "    plt.plot(kneedle.x_normalized, kneedle.y_normalized, label = 'normalized curve', color='blue')\n",
    "    plt.plot(kneedle.x_difference, kneedle.y_difference, label = 'difference curve', color='red')\n",
    "    plt.scatter(kneedle.x_difference[xx], kneedle.y_difference[xx])\n",
    "    plt.text(kneedle.x_difference[xx], kneedle.y_difference[xx]*0.7, 'threshold = '+str(kneedle.x[xx]))\n",
    "    plt.text(np.max(kneedle.x_difference)*0.85/2, 0.70, 'no. of images = '+str(len(images)))\n",
    "    plt.text(np.max(kneedle.x_difference)*0.85/2, 0.65, 'cell number = ') #+str(cellNumber))\n",
    "    plt.text(np.max(kneedle.x_difference)*0.85/2, 0.60, 'imsQ = F') #+str(imsQ))\n",
    "    plt.legend()\n",
    "\n",
    "    plt.subplot(2, 1, 1)\n",
    "    sctx = get_breaking_point(thresholds, count_spots)\n",
    "    plt.plot(thresholds, count_spots)\n",
    "    plt.scatter(sctx[0],count_spots[thresholds==sctx[0]], color='red')\n",
    "    plt.text(sctx[0], count_spots[thresholds==sctx[0]]*1.2, 'threshold = '+str(sctx[0]))\n",
    "    plt.xlabel('Thresholds')\n",
    "    plt.ylabel('mrna detected (log scale)') \n",
    "    return threshold\n",
    "\n",
    "\n",
    "@magicgui(main_window=False,\n",
    "         call_button=\"Verify threshold\",\n",
    "        threshold = {\"widget_type\": \"IntSlider\", 'max': 1000},\n",
    "         )\n",
    "\n",
    "def recordThreshold(img_layer: \"napari.layers.Image\", threshold: int):\n",
    "    image_data = img_layer.data\n",
    "    samples = len(image_data)//10\n",
    "\n",
    "    stackCellSample = image_data[::samples,:,:,:]\n",
    "    if 'stack Cell Sample' not in viewer.layers:\n",
    "        image_layer = viewer.add_image(stackCellSample,\n",
    "                             colormap='green',\n",
    "                                name='stack Cell Sample')\n",
    "\n",
    "    return stackCellSample, threshold\n",
    "\n",
    "\n",
    "    \n",
    "def formatSpotsShape(spotList):\n",
    "    spotArray = np.hstack([np.ones((len(spotList[0]),1))*0, spotList[0]])\n",
    "    for i in range(1,len(spotList)):\n",
    "        spotArray = np.vstack([spotArray,np.hstack([np.ones((len(spotList[i]),1))*i, spotList[i]])])\n",
    "    return spotArray    \n",
    "   \n",
    "\n",
    "@recordThreshold.called.connect\n",
    "def on_verify_threshold(value: str):\n",
    "    print('calling verify threshold')\n",
    "    stackCellSample = value[0]\n",
    "    threshold = value[1]\n",
    "    spots_current_list = []\n",
    "    for t in range(len(stackCellSample)):\n",
    "        rna = np.array(stackCellSample[t])\n",
    "\n",
    "        # spot radiuspathToTimeFrames\n",
    "        spot_radius_px = detection.get_object_radius_pixel(\n",
    "            voxel_size_nm=VOXEL_RADIUS,\n",
    "            object_radius_nm=OBJECT_RADIUS, \n",
    "            ndim=3)\n",
    "\n",
    "        # LoG filter\n",
    "        rna_log = stack.log_filter(rna, sigma=spot_radius_px)\n",
    "\n",
    "        # local maximum detection\n",
    "        mask = detection.local_maximum_detection(rna_log, min_distance=spot_radius_px)\n",
    "\n",
    "        # thresholding\n",
    "\n",
    "        spots_current, _ = detection.spots_thresholding(rna_log, mask, threshold)\n",
    "        spots_current_list.append(spots_current)\n",
    "    spotList = formatSpotsShape(spots_current_list)\n",
    "\n",
    "    bigfish_Spots = viewer.add_points(\n",
    "        spotList,\n",
    "        face_color='#00000000',\n",
    "        size=8,\n",
    "        edge_width=0.3,\n",
    "        edge_width_is_relative=False,\n",
    "        edge_color='white',\n",
    "        name = 'bigFish Detected Spots'+str(threshold)\n",
    "        )\n",
    "    if 'stackCell' in viewer.layers:\n",
    "        viewer.layers.remove('stackCell')\n",
    "@magicgui(main_window=False,\n",
    "         call_button=\"Analyse new cell\",\n",
    "         )\n",
    "def loadNewCell():\n",
    "    viewer.layers.select_all()\n",
    "    viewer.layers.remove_selected()\n",
    "\n",
    "@magicgui(main_window=False,\n",
    "         call_button=\"Analyse new movie\",)\n",
    "def resetParametersForNewMovie():\n",
    "    global CELL_CHOICE\n",
    "    viewer.layers.select_all()\n",
    "    viewer.layers.remove_selected()\n",
    "    choose_home_folder.reset_call_count()\n",
    "    CELL_CHOICE=0\n",
    "    choose_home_folder.dropdown.value = DEFAULT_CHOICES[CELL_CHOICE]    \n",
    "    \n",
    "\n",
    "container = widgets.Container(\n",
    "    widgets=[choose_home_folder, \n",
    "             getFolderName,\n",
    "             showMovie, \n",
    "             setVoxelandObjectParameters, \n",
    "             getThreshold, \n",
    "             recordThreshold, \n",
    "             loadNewCell,resetParametersForNewMovie], layout=\"vertical\", labels=False\n",
    ")\n",
    "container.native.setMinimumWidth(500)\n",
    "\n",
    "viewer = napari.Viewer()\n",
    "viewer.window.add_dock_widget(container, area=\"right\", name = 'threshold estimation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0966c580-b79b-4f98-9770-266d55da5bc2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
